### PR TITLE
Reduce pin related unsafe code

### PR DESCRIPTION
**Pull request by taiki-e**
This removes many of the remaining pin related unsafe code in futures-util and futures-test.

r? @cramertj @Nemo157

### CODE REVIEW

File: futures-test/src/future/assert_unmoved.rs [hunk 1]
```diff
@@ -9,11 +9,14 @@ use std::thread::panicking;
...
 /// method.
+#[pin_project(PinnedDrop)]
 #[derive(Debug, Clone)]
 #[must_use = "futures do nothing unless you `.await` or poll them"]
 pub struct AssertUnmoved<Fut> {
+    #[pin]
     future: Fut,
     this_ptr: *const AssertUnmoved<Fut>,
+    #[pin]
     _pinned: PhantomPinned,
```
Comment (taiki-e) [inline futures-test/src/future/assert_unmoved.rs:20]: Oh, I just added [an option for this in pin-project.](https://docs.rs/pin-project/0.4.14/pin_project/attr.pin_project.html#unpin)

File: futures-util/src/future/future/map.rs [hunk 2]
```diff
@@ -56,34 +44,19 @@ impl<Fut, F, T> Future for Map<Fut, F>
     type Output = T;
 
     #[project]
+    #[project_replace]
```
Comment (taiki-e) [inline futures-util/src/future/future/map.rs:47]: It's not nice to need two wrapper attributes -- I'll fix it soon.

Comment (taiki-e) [reply futures-util/src/future/future/map.rs]: Fixed in pin-project 0.4.15

**Comment by cramertj** [comment]
This looks great, thanks!

**Pull request merged**